fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSession.java

```diff
@@ -0,0 +1,171 @@
+import com.apple.foundationdb.Transaction;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.Tuple;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+@API(API.Status.EXPERIMENTAL)
+public class SynchronizedSession {
+    @Nonnull
+    private Subspace lockSubspace;
+    @Nonnull
+    private FDBDatabaseRunner runner;
+
+    @Nullable
+    private UUID sessionId = null;
+
+    // TODO: configurable?
+    private static final long LEASE_PERIOD_MILL = 60_000;
+
+    private static final Object LOCK_SESSION_ID_KEY = 0L;
+    private static final Object LOCK_SESSION_TIME_KEY = 1L;
+
+    public SynchronizedSession(@Nonnull Subspace lockSubspace, @Nonnull FDBDatabaseRunner runner) {
+        this.lockSubspace = lockSubspace;
+        this.runner = runner;
+    }
+
+    // A synchronized session keeps its look by updating timestamp in each of its transaction, so do not initialize the
+    // session until it is being actively used.
+    public SynchronizedSessionRunner initializeSessionAndCreateRunner() {
+        boolean success = runner.run(context -> runner.asyncToSync(FDBStoreTimer.Waits.WAIT_INITIALIZE_SYNCHRONIZED_SESSION, initializeSession(context)));
+        if (success) {
+            return new SynchronizedSessionRunner(runner, this);
+        } else {
+            throw new SynchronizedSessionExpiredException("failed to initialize the session");
+        }
+    }
+
+    public SynchronizedSessionRunner reuseSessionAndCreateRunner(@Nonnull UUID sessionId) {
+        this.sessionId = sessionId;
+        return new SynchronizedSessionRunner(runner, this);
+    }
+
+    // Return true if get lease otherwise false.
+    private CompletableFuture<Boolean> initializeSession(@Nonnull FDBRecordContext context) {
+        if (sessionId != null) {
+            throw new RecordCoreException("SynchronizedSession has been initialized");
+        }
+        sessionId = UUID.randomUUID();
+        Transaction tr = context.ensureActive();
+        return getLockSessionId(tr).thenCompose(lockSessionId -> {
+            if (lockSessionId == null) {
+                // If there was no lock, can get the lock.
+                return getSessionLock(tr);
+            } else if (lockSessionId == sessionId) {
+                // This should never happen.
+                throw new RecordCoreException("session id already exists");
```

Comment:
in the final code, this should include the subspace and id.

Reply:
Done.